feat(search page): create pagination and card for search page [FT-05]

anindyapramudita opened the pull request:
### PR SUMMARY:
**WHAT**
Create a complete functioning search page

**WHY**
So user can filter through the database to find the perfect plant according to their needs and preference

**HOW**
- Create card component to show the filtered plant data
- Integrate the card data with the functionality of the filter
- Create and integrate pagination component

**This PR Consist of:**
[ ] New dependencies
[ ] Version update in dependencies
[ ] Changes in .env file

**Documentation:**
![image](https://github.com/anindyapramudita/cacti-plant-website/assets/97867884/da1173f9-c383-401e-b7c2-f961f9dd75fd)
![image](https://github.com/anindyapramudita/cacti-plant-website/assets/97867884/892919fb-fe10-4045-9d94-59f34bf9cd25)

**Video link:** https://we.tl/t-PseuPw9St4

**Ticket Link:** https://www.notion.so/FT-05-Create-cards-for-filter-page-0e81c31bfa6d47f8b57bdcdea42f9770?pvs=4

red-runner commented on src/components/layouts/small-card/small-card-layout.styles.tsx:
```diff
@@ -0,0 +1,34 @@
+  gap: var(--gap-size);
+  padding: var(--padding-size);
+
+  @media (min-width: ${gridBreakpoint.sm}) {
```
Remember to delete 🦆

anindyapramudita replied:
🥔 Done!

red-runner commented on src/components/pagination/pagination.styles.tsx:
```diff
@@ -0,0 +1,26 @@
+import styled from "styled-components";
+
+export const StylesWrapper = styled.div<{}>`
```
Remember you can declare css variables with values and consume them in the CSS 🦆

anindyapramudita replied:
🥔 Done! I've updated the styles file and use css variables

red-runner commented on src/components/pagination/pagination.tsx:
```diff
@@ -0,0 +1,76 @@
+    return null;
+  }
+
+  const getClassName = (id: number) => {
```
To avoid developing a function each time for calculating the classname. You can install the classnames npm package: 
https://www.npmjs.com/package/classnames. This example would look like this:

```
classnames('page-button',  {
  'active': id + 1 === currentPage,
})
```

anindyapramudita replied:
🥔 I've installed this package and it worked perfectly! It's included in the latest push :)

red-runner commented on src/components/pagination/pagination.tsx:
```diff
@@ -0,0 +1,76 @@
+    return "page-button";
+  };
+
+  const getPagesValidity = (id: number) =>
```
With utilities, it is recommended to separate them into a utils folder. If the utils are specific for this component, you can create a utils folder locally in the component folder. Else, if the function is meant to be used globally, you can expose it in a global utils folder :) This also provides further advantages, like, functions being scoped and pure, which allows you to create tests specifically to test the results of that function. :)

anindyapramudita replied:
🥔 noted, now this is moved to a utils folder specifically for the pagination component

red-runner commented on src/components/pagination/pagination.tsx:
```diff
@@ -0,0 +1,76 @@
+    (currentPage + 2 > totalPage && totalPage - (id + 1) < 5) || // to check if current page is 1 less than the last page
+    Math.abs(id + 1 - currentPage) <= 2; // to only print 2 pages before and after current page
+
+  const getPages = (id: number) => {
```
For functions that are meant to output JSX, you can use the prefix `render-` --> `renderPages()`

anindyapramudita replied:
🥔 donezies as well!

red-runner commented on src/components/small-card/small-card.tsx:
```diff
@@ -0,0 +1,57 @@
+}) => {
+  const [isLiked, setIsLiked] = useState<boolean>(false);
+
+  const handleGetLevel = (level: number) => {
```
For handling this kind of key value situations, you can create an object in your constants file, that behaves like a dictionary for example in this case, we can do:

```
const levels = {
  1: 'Low',
  2: 'Medium',
  3: 'High'
}
```

Then in your function you can do:

```
const handleGetLevel = (level = 1) => {
  return levels[number];
```

You can also set a default, so the function always returns a value inside the dictionary, in case the level is undefined :)

anindyapramudita replied:
🥔 I've implemented this suggestion as well :)

red-runner commented on src/components/small-card/small-card.tsx:
```diff
@@ -0,0 +1,57 @@
+      <div className="card-image">
+        <Link href={`/plant/${data._id}`}>
+          <TransitionImage
+            image={data.images[0].src}
```
Hi! Be aware that if images[0] does not have an entry, this will lead to uncaught error :) You can either set a default, check with an if statement, or use chain operator like so: `data?.images?.[0]?.src`

anindyapramudita replied:
🥔 Just to be safe, I did both. I updated to use the chain operator like so and also added default value to the image component if the image is not found / undefined

anindyapramudita commented:
> Some stuff to fix, but looks very nice :) One thing to realize from the pictures, is that the pagination, follows the images. If there number of images change while users are searching, this might lead to the pagination to bounce up and down during the search. For a better experience, you might want to make it stay at the bottom of the page :)

🥔 I've done this as well and made sure the pagination is always at the bottom, even if there is only one line of cards that shows up :)

anindyapramudita commented:
@red-runner henlo! I updated this today by refactoring the components that are used in this feature such as the filter accordion, filter header, and pagination. Aside from that, I also created tests and added them to the component documentation with Storybook. Please kindly check them as well when you review this PR hehe 🥔🧐
